File: ai/zero/selfplay.py
"""
Generating training data by playing the network against itself.

Each position played produces one example: the board as its mover saw it, the search's visit
counts as the policy to imitate, and - once the game is finished - what the game was worth to
that mover. The network is then trained to predict, in one position, what the search concluded
after looking ahead from it. That is the whole trick: search is slow and better than the network,
so the network chases it, and the better network makes the next search better still.

Two things about the value target are easy to get wrong and silent when wrong:

* **It is from the mover's own point of view.** A position whose player to move went on to win is
  `+1` for that position, whichever player it was.
* **A draw is 0, not a sign.** Most tic-tac-toe self-play games end level and most examples should
  say so.
"""

import random
from typing import (
    Any,
    Callable,
    Iterator,
    List,
    NamedTuple,
    Optional,
    Sequence,
    Tuple,
)

from games.base import Encoder, GameState
from ai.zero.mcts import (
    DIRICHLET_ALPHA,
    DIRICHLET_EPSILON,
    EXPLORATION,
    MCTS,
    Evaluator,
    drive,
)

# Plies played by sampling from the visit counts rather than taking the best move, without which
# every self-play game from a deterministic network is the same game. AlphaZero's number, which
# for tic-tac-toe covers the whole game: cutting it to three plies cost 10 points of agreement.
TEMPERATURE_MOVES = 30
TEMPERATURE = 1.0

# What the temperature drops to afterwards. Zero is greedy - more aggressive than AlphaZero, which
# keeps sampling at a low temperature. A game whose self-play needs late diversity can raise it.
FINAL_TEMPERATURE = 0.0

# Random plies played before a game starts being recorded, drawn uniformly from 0 to this.
#
# Off, because exploration belongs in PUCT and root noise: random starts here reached the same
# agreement as raising c_puct to 5.0 did, and needing them is a sign that one of those is mistuned.
# Kept for benchmarks, which want coverage of positions real play never reaches.
OPENING_PLIES = 0


class Example(NamedTuple):
    """One position, and what the search and the eventual result had to say about it."""

    planes: Any
    policy: Sequence[float]
    value: float


def game_steps(
    encoder: Encoder,
    game: Callable[[], GameState],
    simulations: int,
    temperature_moves: int = TEMPERATURE_MOVES,
    temperature: float = TEMPERATURE,
    final_temperature: float = FINAL_TEMPERATURE,
    opening_plies: int = 0,
    exploration: float = EXPLORATION,
    dirichlet_alpha: float = DIRICHLET_ALPHA,
    dirichlet_epsilon: float = DIRICHLET_EPSILON,
    rng: Optional[random.Random] = None,
) -> Iterator[GameState]:
    """
    A whole game as a generator: it yields every position it needs evaluated and returns
    `(examples, finished state)`.

    Suspendable for the same reason the search is, and `yield from` is all it takes to stay so.

    The finished state comes back so a caller can report on the games as well as learn from them.

    `opening_plies` puts the game somewhere random before any of it is recorded, so the randomness
    moves where games happen without changing what is learned from them.
    """
    rng = rng or random.Random()
    mcts = MCTS(
        _refuses, encoder, simulations=simulations, exploration=exploration,
        dirichlet_alpha=dirichlet_alpha, dirichlet_epsilon=dirichlet_epsilon, rng=rng,
    )

    state = _opening(game, opening_plies, rng)
    played: List[Tuple[Any, Sequence[float], bool]] = []

    while not state.is_game_over:
        result = yield from mcts.steps(state, noise=True)
        tau = temperature if len(played) < temperature_moves else final_temperature

        # Recorded before the move: the example belongs to the position it was searched from.
        played.append((encoder.planes(state), result.policy, state.turn))
        state.make_move(mcts.sample(result.visits, tau))

    outcome = state.result
    examples = [
        Example(planes, policy, _value_to(mover, outcome.winner))
        for planes, policy, mover in played
    ]
    return examples, state


def _refuses(state: GameState):
    """
    The evaluator an `MCTS` built for `game_steps` will never call.

    `MCTS.steps` hands every request to whoever drives it, which here is the batched caller. This
    raises so that a stray `search()` fails loudly rather than quietly using a different evaluator.
    """
    raise RuntimeError('this search is driven by its caller and has no evaluator of its own')


def play_game(
    evaluator: Evaluator,
    encoder: Encoder,
    game: Callable[[], GameState],
    simulations: int,
    **kwargs,
) -> Tuple[List[Example], GameState]:
    """One game, answering the search one position at a time. `play_games` is the batched form."""
    return drive(game_steps(encoder, game, simulations, **kwargs), evaluator)


def play_games(
    batch_evaluator: Callable[[Sequence[GameState]], Sequence[Tuple[Sequence[float], float]]],
    encoder: Encoder,
    game: Callable[[], GameState],
    count: int,
    simulations: int,
    batch_size: int = 1,
    seed: int = 0,
    on_finished: Optional[Callable[[int, int], None]] = None,
    **kwargs,
) -> List[Tuple[List[Example], GameState]]:
    """
    Plays `count` games, keeping up to `batch_size` of them in flight and evaluating the positions
    they are waiting on together.

    This is where most of a run's time goes and where it is won back: the search asks for one
    position per simulation, and a batch of sixty-four is ten times the throughput of a batch of
    one.

    **Each game keeps its own random stream**, seeded from `seed` and the game's index, so
    `batch_size` cannot change what is played. Games sharing a stream would draw from it in
    whatever order they happened to reach it.

    The evaluator is handed every pending position at once and must read what it needs from all of
    them before returning: the states are live and resuming a game mutates its own.

    `on_finished(completed, count)` is called as each game ends, which is the only sign of life a
    twenty-minute Connect 4 generation gives.
    """
    started, done = 0, {}
    active: List[Tuple[int, Iterator[GameState], GameState]] = []

    while started < count or active:
        while started < count and len(active) < batch_size:
            steps = game_steps(
                encoder, game, simulations, rng=random.Random(f'{seed}:{started}'), **kwargs)
            _advance(started, steps, None, active, done, on_finished, count)
            started += 1

        if not active:
            break

        # Every pending position, answered in one pass, before any of them is resumed.
        answers = batch_evaluator([pending for _, _, pending in active])
        waiting, active = active, []
        for (index, steps, _), answer in zip(waiting, answers):
            _advance(index, steps, answer, active, done, on_finished, count)

    # By game index rather than completion order, which is a function of `batch_size`.
    return [done[index] for index in range(count)]


def _advance(index, steps, answer, active, done, on_finished=None, count=0) -> None:
    """Pushes one game forward, onto `active` if it wants another position or `done` if finished."""
    try:
        active.append((index, steps, steps.send(answer)))
    except StopIteration as finished:
        done[index] = finished.value
        if on_finished:
            on_finished(len(done), count)


def _opening(game: Callable[[], GameState], plies: int, rng: random.Random) -> GameState:
    """
    A position a random number of random plies in, and still running.

    The count is drawn per game so openings land at every depth rather than piling up at one. A
    game that finishes inside its own opening is redrawn.
    """
    if plies <= 0:
        return game()

    for _ in range(50):
        state = game()
        for _ in range(rng.randint(0, plies)):
            if state.is_game_over:
                break
            state.make_move(rng.choice(list(state.legal_moves)))

        if not state.is_game_over:
            return state
    return game()


def _value_to(mover: bool, winner: Optional[bool]) -> float:
    """What a finished game was worth to the player who was about to move."""
    if winner is None:
        return 0.0
    return 1.0 if winner == mover else -1.0


def augment(examples: Sequence[Example], encoder: Encoder) -> List[Example]:
    """
    Every example under every symmetry the game admits.

    The value is carried through untouched: rotating a board does not change who is winning.
    """
    grown: List[Example] = []
    for example in examples:
        for planes, policy in encoder.symmetries(example.planes, list(example.policy)):
            grown.append(Example(planes, policy, example.value))
    return grown

File: rust/crates/c4-core/src/selfplay.rs
//! Many self-play games advanced together, so the positions they wait on can be evaluated at once.
//!
//! A port of `ai/zero/selfplay.py`. The shape is the one that module already argues for: one
//! tree's simulations cannot be batched, because each goes where the previous ones' statistics
//! send it, so the batching happens *between* games instead and every game gets the tree it would
//! have had alone. The only change here is the scale it can be run at - a Python batch is capped
//! by how many trees Python can afford to walk, and this one is capped by nothing much, which is
//! what lets a whole generation be in flight and a GPU be fed a batch it can use.
//!
//! Each game keeps its own random stream, seeded from the run's seed and the game's index, so how
//! many are in flight cannot change what is played.

use crate::connect4::Connect4;
use crate::encode::{legal_mask, planes, PLANE_BYTES, POLICY_SIZE};
use crate::mcts::{Search, State};
use crate::rng::PyRandom;

pub struct Config {
    pub games: usize,
    pub in_flight: usize,
    pub simulations: u32,
    pub exploration: f64,
    pub dirichlet_alpha: f64,
    pub dirichlet_epsilon: f64,
    pub temperature: f64,
    pub final_temperature: f64,
    pub temperature_moves: usize,
    pub opening_plies: u32,
    pub seed: String,
}

/// One finished game's training examples, and how it ended.
pub struct Record {
    pub planes: Vec<u8>,
    pub policies: Vec<f64>,
    pub values: Vec<f64>,
    pub positions: usize,
    pub winner: Option<bool>,
}

/// The positions every game in flight is waiting on, in slot order.
pub struct Batch<'a> {
    pub planes: &'a [u8],
    pub legal: &'a [bool],
    pub positions: usize,
}

struct Game {
    index: usize,
    board: Connect4,
    search: Search,
    rng: PyRandom,
    planes: Vec<u8>,
    policies: Vec<f64>,
    movers: Vec<bool>,
    done: bool,
}

impl Game {
    fn start(index: usize, config: &Config) -> Self {
        let mut rng = PyRandom::from_text(&format!("{}:{}", config.seed, index));
        let board = opening(config.opening_plies, &mut rng);
        let done = board.is_game_over();

        let mut game = Game {
            index,
            board,
            search: Search::new(
                config.simulations,
                config.exploration,
                config.dirichlet_alpha,
                config.dirichlet_epsilon,
                true,
            ),
            rng,
            planes: Vec::new(),
            policies: Vec::new(),
            movers: Vec::new(),
            done,
        };
        if !game.done {
            game.search.begin();
        }
        game
    }

    /// Answers the position this game is waiting on, and plays a move if that finished a search.
    fn advance(&mut self, priors: &[f32], value: f32, config: &Config) {
        let mut widened = [0.0f64; POLICY_SIZE];
        for (slot, prior) in priors.iter().enumerate() {
            widened[slot] = *prior as f64;
        }
        self.search
            .submit(&mut self.board, &widened, value as f64, &mut self.rng);

        if *self.search.state() != State::Finished {
            return;
        }

        let result = self.search.result();

        // Recorded before the move: the example belongs to the position it was searched from.
        let start = self.planes.len();
        self.planes.resize(start + PLANE_BYTES, 0);
        planes(&self.board, &mut self.planes[start..]);
        self.policies.extend_from_slice(&result.policy());
        self.movers.push(self.board.turn());

        let temperature = if self.movers.len() <= config.temperature_moves {
            config.temperature
        } else {
            config.final_temperature
        };
        self.board.make_move(result.sample(temperature, &mut self.rng));

        if self.board.is_game_over() {
            self.done = true;
        } else {
            self.search.begin();
        }
    }

    fn finish(self) -> Record {
        let winner = self.board.result().expect("the game is not over").winner;
        let values = self
            .movers
            .iter()
            .map(|&mover| value_to(mover, winner))
            .collect();

        Record {
            planes: self.planes,
            policies: self.policies,
            values,
            positions: self.movers.len(),
            winner,
        }
    }
}

pub struct SelfPlay {
    config: Config,
    active: Vec<Game>,
    started: usize,
    completed: usize,
    records: Vec<Option<Record>>,
    planes: Vec<u8>,
    legal: Vec<bool>,
}

impl SelfPlay {
    pub fn new(config: Config) -> Self {
        let in_flight = config.in_flight.clamp(1, config.games.max(1));
        SelfPlay {
            active: Vec::with_capacity(in_flight),
            started: 0,
            completed: 0,
            records: (0..config.games).map(|_| None).collect(),
            planes: Vec::with_capacity(in_flight * PLANE_BYTES),
            legal: Vec::with_capacity(in_flight * POLICY_SIZE),
            config: Config { in_flight, ..config },
        }
    }

    pub fn completed(&self) -> usize {
        self.completed
    }

    /// Starts games until the flight is full, then reports what they are all waiting on.
    ///
    /// `None` means every game has finished. The caller must evaluate the whole batch and hand it
    /// back through `submit` before asking again.
    pub fn pending(&mut self) -> Option<Batch<'_>> {
        while self.started < self.config.games && self.active.len() < self.config.in_flight {
            let game = Game::start(self.started, &self.config);
            self.started += 1;
            if game.done {
                self.retire(game);
            } else {
                self.active.push(game);
            }
        }

        if self.active.is_empty() {
            return None;
        }

        self.planes.clear();
        self.planes.resize(self.active.len() * PLANE_BYTES, 0);
        self.legal.clear();
        self.legal.resize(self.active.len() * POLICY_SIZE, false);

        for (slot, game) in self.active.iter().enumerate() {
            planes(
                &game.board,
                &mut self.planes[slot * PLANE_BYTES..(slot + 1) * PLANE_BYTES],
            );
            legal_mask(
                &game.board,
                &mut self.legal[slot * POLICY_SIZE..(slot + 1) * POLICY_SIZE],
            );
        }

        Some(Batch {
            planes: &self.planes,
            legal: &self.legal,
            positions: self.active.len(),
        })
    }

    /// Answers the batch `pending` reported, in the same order.
    ///
    /// Single-threaded on purpose. The games are independent and could be advanced across
    /// threads, but one game's share of a batch is a few hundred nanoseconds against tens of
    /// thousands of batches, so the synchronisation costs several times what the work does. The
    /// parallelism that pays here is the batch itself, and that happens on the GPU.
    pub fn submit(&mut self, priors: &[f32], values: &[f32]) {
        assert_eq!(priors.len(), self.active.len() * POLICY_SIZE);
        assert_eq!(values.len(), self.active.len());

        let config = &self.config;
        self.active
            .iter_mut()
            .zip(priors.chunks(POLICY_SIZE))
            .zip(values.iter())
            .for_each(|((game, priors), &value)| game.advance(priors, value, config));

        let mut slot = 0;
        while slot < self.active.len() {
            if self.active[slot].done {
                let game = self.active.swap_remove(slot);
                self.retire(game);
            } else {
                slot += 1;
            }
        }
    }

    fn retire(&mut self, game: Game) {
        let index = game.index;
        self.records[index] = Some(game.finish());
        self.completed += 1;
    }

    /// Every game's examples, in game order rather than the order they happened to finish in.
    pub fn records(&self) -> impl Iterator<Item = &Record> {
        self.records.iter().map(|record| {
            record
                .as_ref()
                .expect("a game has not finished; drive `pending` to `None` first")
        })
    }

    pub fn positions(&self) -> usize {
        self.records
            .iter()
            .filter_map(|record| record.as_ref())
            .map(|record| record.positions)
            .sum()
    }
}

/// A position a random number of random plies in, and still running. Zero plies - the measured
/// default - is the empty board and draws nothing from the stream.
fn opening(plies: u32, rng: &mut PyRandom) -> Connect4 {
    if plies == 0 {
        return Connect4::new();
    }

    let mut moves = [0u8; POLICY_SIZE];
    for _ in 0..50 {
        let mut board = Connect4::new();
        for _ in 0..rng.below(plies + 1) {
            if board.is_game_over() {
                break;
            }
            let count = board.legal_moves(&mut moves);
            board.make_move(moves[rng.below(count as u32) as usize]);
        }

        if !board.is_game_over() {
            return board;
        }
    }
    Connect4::new()
}

/// What a finished game was worth to the player who was about to move.
fn value_to(mover: bool, winner: Option<bool>) -> f64 {
    match winner {
        None => 0.0,
        Some(winner) => {
            if winner == mover {
                1.0
            } else {
                -1.0
            }
        }
    }
}
